testing/unittests/jthreadtests.cpp

```diff
@@ -0,0 +1,591 @@
+                                                    E->errorMessage(msg);
+                                                    CPPUNIT_ASSERT_EQUAL_STR("No threads available in pool ThrottledTestPoolWithFastThreadCompletion", msg.str());
+                                                    ++exceptionCount;
+                                                    E->Release();
+                                                }
+
+                                                // Record total time spent waiting (includes throttle delays and timeouts)
+                                                auto afterDelayedStart = std::chrono::high_resolution_clock::now();
+                                                auto startDuration = std::chrono::duration_cast<std::chrono::milliseconds>(afterDelayedStart - beforeDelayedStart);
+                                                totalStartDuration += startDuration.count(); });
+
+                asyncStart.detach();
+            }
+
+            // Wait 5 seconds for all async operations to complete
+            std::this_thread::sleep_for(std::chrono::seconds(5));
+
+            CPPUNIT_ASSERT(totalStartDuration.load() >= 10000LL && totalStartDuration.load() <= 10001LL);
+
+            CPPUNIT_ASSERT_EQUAL(0U, pool->runningCount());          // Pool drained
+            CPPUNIT_ASSERT_EQUAL(300U, threadStartedCount.load());   // 200 from phase 1 + ~100 from phase 2
+            CPPUNIT_ASSERT_EQUAL(300U, threadCompletedCount.load()); // All started threads completed
+
+            CPPUNIT_ASSERT(exceptionCount >= 99U && exceptionCount <= 101U);
+        }
+    }
+
+    void testWaitAvailable()
+    {
+        // Test: Pool with unlimited capacity should always return true
+        pool.setown(createThreadPool("UnlimitedTestPool", factory, true, nullptr, 0)); // defaultmax=0 means unlimited
+
+        CPPUNIT_ASSERT(pool->waitAvailable(100));
+        CPPUNIT_ASSERT(pool->waitAvailable(0));
+
+        pool.clear();
+
+        // Test: Pool with limited capacity
+        pool.setown(createThreadPool("LimitedTestPool", factory, true, nullptr, 2)); // max 2 threads
+
+        // Should return true when pool is empty
+        CPPUNIT_ASSERT(pool->waitAvailable(100));
+
+        pool->start(&lifespan5seconds, "Thread1");
+        pool->start(&lifespan5seconds, "Thread2");
+
+        std::this_thread::sleep_for(std::chrono::milliseconds(50));
+        CPPUNIT_ASSERT_EQUAL(2U, pool->runningCount());
+
+        // Test: waitAvailable should return false when pool is full and timeout expires
+        auto startTime = std::chrono::high_resolution_clock::now();
+        CPPUNIT_ASSERT(!pool->waitAvailable(200));
+        auto endTime = std::chrono::high_resolution_clock::now();
+        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime);
+
+        CPPUNIT_ASSERT(duration.count() >= 180 && duration.count() <= 250);
```

Comment:
why 180? and is 50ms margin enough (should us esignals and failure on longer than expected signalling )